### PR TITLE
Ecosystem import job

### PR DESCRIPTION
**Pull request by karenc**
- Use background job for admin ecosystem importing textbook

  Use `Marshal.dump` when passing in the tag_generator to
  `FetchAndImportBookAndCreateEcosystem`.  An instance of the concept
  coach tag generator cannot be used in background jobs:

  ```Unsupported argument type: ConceptCoach::TagGenerator```

- Store the book url with the ecosystem import background job

- Display incomplete ecosystem import jobs on admin ecosystems page

  Display job id (with link to the job details page), job status and link
  to the book in archive.

- Change ecosystems index page to use tabs

  "List of Books" and "Incomplete Imports".

  Add note to "Incomplete Imports" table to say that the table does not
  automatically refresh.

### CODE REVIEW

File: app/views/manager/ecosystems/_incomplete.html.erb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<h2>Incomplete Ecosystem Imports</h2>
+
```
Comment (jpslav) [inline app/views/manager/ecosystems/_incomplete.html.erb]: Can you add a note that says that this table does not automatically refresh (possible that some might expect it to)

File: app/controllers/admin/ecosystems_controller.rb [hunk 2]
```diff
@@ -14,11 +17,14 @@ def import_ecosystem
     archive_url = params[:archive_url].present? ? params[:archive_url] : @default_archive_url
 
     OpenStax::Cnx::V1.with_archive_url(url: archive_url) do
-      ecosystem = FetchAndImportBookAndCreateEcosystem[
+      job_id = FetchAndImportBookAndCreateEcosystem.perform_later(
         book_cnx_id: params[:cnx_id],
-        tag_generator: ConceptCoach::TagGenerator.new(params[:cc_tag])
```
Comment (jpslav) [inline app/controllers/admin/ecosystems_controller.rb]: I know why you are passing in the `cc_tag` here instead of the `tag_generator`, but I would very much like to keep the knowledge of concept coach stuff at this level of the code and not inside the Content subsystem.

We can do this by marhsalling the `tag_generator` to a string and then unmarshalling it inside the routine:

```ruby
# EcosystemsController
job_id = FetchAndImportBookAndCreateEcosystem.perform_later(
  ...
  tag_generator: Marshal.dump(ConceptCoach::TagGenerator.new(params[:cc_tag]))
)
```

```ruby
# FetchAndImportBookAndCreateEcosystem
def exec(..., tag_generator: nil)
  ...
  tag_generator = Marshal.load(tag_generator) if tag_generator.is_a?(String)
```

We should make marshalling and unmarshalling standard procedure for any routine that we might call in a background job that takes non-primitive, non-activerecord arguments.

File: app/routines/fetch_and_import_book_and_create_ecosystem.rb [hunk 3]
```diff
@@ -8,10 +8,11 @@ class FetchAndImportBookAndCreateEcosystem
   protected
 
   # Returns a Content::Ecosystem containing a book obtained from the given CNX id
-  def exec(book_cnx_id:, ecosystem_title: nil, exercise_uids: nil, tag_generator: nil)
+  def exec(book_cnx_id:, ecosystem_title: nil, exercise_uids: nil, tag_generator: nil, cc_tag: nil)
     cnx_book = OpenStax::Cnx::V1.book(id: book_cnx_id)
     eco_title ||= "#{cnx_book.title} (#{cnx_book.uuid}@#{cnx_book.version}) - #{Time.now.utc}"
     outputs[:ecosystem] = Content::Ecosystem.create!(title: eco_title)
+    tag_generator ||= ConceptCoach::TagGenerator.new(cc_tag) unless cc_tag.nil?
```
Comment (jpslav) [inline app/routines/fetch_and_import_book_and_create_ecosystem.rb]: See comment above.  Also, don't need `||=` here.

**Comment by jpslav** [comment]
Thanks @karenc! Looks great.  Just a few comments.  Oh, and does Lev's automatic reporting of errors to the background job handle the "get any errors out to the admins" issue?

File: app/controllers/admin/ecosystems_controller.rb (hunk 2 above)
Comment (Dantemss) [inline app/controllers/admin/ecosystems_controller.rb]: Wait... why is this needed at all? Is ActiveJob unable to automatically serialize non-ActiveRecord stuff?

**Comment by karenc** [comment]
Added note to say the incomplete imports table does not automatically refresh.

Also changed the ecosystems index page to use tabs:

ecosystems1

ecosystems2

File: app/controllers/admin/ecosystems_controller.rb (hunk 2 above)
Comment (karenc) [inline app/controllers/admin/ecosystems_controller.rb]: Yeah I got this error: `Unsupported argument type: ConceptCoach::TagGenerator`

**Comment by karenc** [comment]
The error isn't really clear... I put in an invalid uuid for the cnx book id and the error on the job details page is "exception - Can't convert NilClass into String.".

@jpslav Do you have an example book that has errors in the content?

May be I can add a link to papertrail (all systems) that searches for the error message on the job details page?

**Comment by jpslav** [comment]
@karenc I think if the routine actually added an error instead of just raising an exception (which it does in places), the message will be more meaningful.  I don't have an example of erroring book off hand.  If we really need it I bet we could find an problematic version of a book, but I think we can pass for the moment.

File: app/controllers/admin/ecosystems_controller.rb (hunk 2 above)
Comment (jpslav) [inline app/controllers/admin/ecosystems_controller.rb]: I don't think so @Dantemss -- ActiveJob only supports serialization of ActiveRecords, and only then because they support GlobalID.
